python/src/spark_rapids_ml/dbscan.py

```diff
@@ -0,0 +1,747 @@
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = _CumlCaller._pre_process_data(self, dataset)
+
+        # Must retain idCol for label matching
+        if self.hasParam("idCol") and self.isDefined("idCol"):
+            id_col_name = self.getOrDefault("idCol")
+            select_cols.append(col(id_col_name).alias(alias.row_number))
+        else:
+            select_cols.append(col(alias.row_number))
+
+        return select_cols, multi_col_names, dimension, feature_type
+
+    def _out_schema(
+        self, input_schema: StructType = StructType()
+    ) -> Union[StructType, str]:
+        return StructType(
+            [
+                StructField(self._get_prediction_name(), IntegerType(), False),
+                StructField(self.getIdCol(), LongType(), False),
+            ]
+        )
+
+    def _transform_array_order(self) -> _ArrayOrder:
+        return "C"
+
+    def _fit_array_order(self) -> _ArrayOrder:
+        return "C"
+
+    def _require_nccl_ucx(self) -> Tuple[bool, bool]:
+        return (True, True)
+
+    def _get_cuml_fit_func(
+        self,
+        dataset: DataFrame,
+        extra_params: Optional[List[Dict[str, Any]]] = None,
+    ) -> Callable[
+        [FitInputType, Dict[str, Any]],
+        Dict[str, Any],
+    ]:
+        import cupy as cp
+        import cupyx
+
+        dtype = self.dtype
+        n_cols = self.n_cols
+        array_order = self._fit_array_order()
+        pred_name = self._get_prediction_name()
+        idCol_name = self.getIdCol()
+        logger = get_logger(self.__class__)
+
+        cuda_managed_mem_enabled = (
+            _get_spark_session().conf.get("spark.rapids.ml.uvm.enabled", "false")
+            == "true"
+        )
+
+        inputs = []  # type: ignore
+
+        idCol = list(
+            self.idCols_[0].value
+            if len(self.idCols_) == 1
+            else np.concatenate([chunk.value for chunk in self.idCols_])
+        )
+
+        for pdf_bc in self.raw_data_:
+            pdf = pd.DataFrame(data=pdf_bc.value, columns=self.processed_input_cols)
+
+            if self.multi_col_names:
+                features = np.array(pdf[self.multi_col_names], order=array_order)
+            elif self.use_sparse_array:
+                # sparse vector
+                features = _read_csr_matrix_from_unwrapped_spark_vec(pdf)
+            else:
+                # dense vector
+                features = np.array(list(pdf[alias.data]), order=array_order)
+
+            # experiments indicate it is faster to convert to numpy array and then to cupy array than directly
+            # invoking cupy array on the list
+            if cuda_managed_mem_enabled:
+                features = (
+                    cp.array(features)
+                    if self.use_sparse_array is False
+                    else cupyx.scipy.sparse.csr_matrix(features)
+                )
+
+            inputs.append(features)
```

Comment:
code snippet for preparing inputs variable seemed very similar to that of core.py. Is it possible to reuse the code snippet of core.py? 

Seems you only need the concated variable in _cuml_fit, while you should be able to use dfs variable to construct concated.

Reply:
The issue here is that we need to work on the original full dataset instead of the dfs passed into _cuml_fit which is the partitioned dataframe for mapInPandas. There is no way to reconstruct from dfs the full dataset. The code is iterating on the broadcast varibales. The code snippet is partially duplicated due to it's a subportion of code from core.py -- _call_cuml_fit_func, while we have to extract this portion of code to work on a different set of broadcast variables (not the passed data to core.py).